File: algorithms/shunting_yard.java
// The Shunting Yard algorithm is used to convert infix expressions (where operators are placed between operands) into postfix notation (operators follow their operands).
// It uses a stack to temporarily hold operators and performs a series of operations based on the precedence and associativity of operators.

// Imagine you're designing a calculator application where users can input mathematical expressions containing operators like `+`, `-`, `*`, `/`, and parantheses `(`,`)`.
// The Shunting Yard algorithm can be used to convert these infix expressions into postfix (Reverse Polish notation - RPN) format, which is easier to evaluate using a stack-based approach.

import java.util.*;

public class ShuntingYard {

    public static String infixToPostfix(String infixExpression) {
        StringBuilder postfix = new StringBuilder();
        Deque<Character> operators = new ArrayDeque<>();

        Map<Character, Integer> precedence = Map.of("+", 1, "-", 1, "*", 2, "/", 2);

        for (char c : infixExpression.toCharArray()) {
            if (Character.isDigit(c)) {
                postfix.append(c); // Operand directly added to output
            } else if (c == "(") {
                operators.push(c); // Left paranthesis pushed onto stack
            } else if (c == ")") {
                // Pop operators from stack to output until '(' is encountered
                while (!operators.isEmpty() && operators.peek() != "(") {
                    postfix.append(operators.pop());
                }
                operators.pop(); // Remove "(" from stack
            } else {
                // Token is an operator
                while (!operators.isEmpty() && precedence.getOrDefault(c, 0) <= precedence.getOrDefault(operators.peek(), 0)) {
                    postfix.append(operators.pop());
                }
                operators.push(c); // Push current operator onto stack
            }
        }

        // Pop remaining operators from stack to output
        while (!operators.isEmpty()) {
            postfix.append(operators.pop());
        }

        return postfix.toString();
    }

    public static void main(String[] args) {
        String infixExpression = "3 + 4 * (2 - 1)";
        String postfixExpression = infixToPostfix(infixExpression);
        System.out.println("Infix: " + infixExpression);
        System.out.println("Postfix: " + postfixExpression);
    }
}

File: algorithms/shunting_yard.kt
// The Shunting Yard Algorithm, devised by Edsger Dijkstra, is used to parse mathematical expressions specified in infix notation (e.g., "3 + 4") and produce an equivalent postfix notation expression (e.g., "3 4 +").
// This postfix notation is easier to evaluate algorithmically without the need for precedence rules or parantheses.

// Imagine you're developing a calculator app that needs to handle complex mathematical expressions input by the user, such as "3 + 4 * 2 / (1 - 5) ^ 2 ^ 3".
// The challenge is to correctly interpret the order of operations, including paranthesis.
// The Shunting Yard Algorithm is used to convert expressions from infix notation (which is human-readable) to postfix notation (which is easier for a computer to evaluate).

fun shuntingYard(expression: String): String {
    val outputQueue = ArrayDeque<String>()
    val operatorStack = ArrayDeque<Char>()
    val precedence = mapOf("+" to 1, "-" to 1, "*" to 2, "/" to 2, "^" to 3)

    val tokens = expression.replace(Regex("\\s+"), "").split(Regex("(?<=[-+*/^()])|(?=[-+*/^()])"))

    for (token in tokens) {
        when {
            token.isDouble() -> outputQueue.addLast(token)
            token.first() in precedence.keys -> {
                while (operatorStack.isnotEmpty() && precedence[operatorStack.last()]!! >= precedence[token.first()]!!) {
                    outputQueue.addLast(operatorStack.removeLast().toString())
                }
                operatorStack.addLast(token.first())
            }
            token.first() == "(" -> operatorStack.addLast(token.first())
            token.first() == ")" -> {
                while (operatorStack.last() != "(") {
                    outputQueue.addLast(operatorStack.removeLast().toString())
                }
                operatorStack.removeLast() // Remove the "(" from the stack
            }
        }
    }

    while (operatorStack.isNotEmpty()) {
        outputQueue.addLast(operatorStack.removeLast().toString())
    }

    return outputQueue.joinToString(" ")
}

fun String.isDouble(): Boolean = this.toDoubleOrNull() != null

fun main() {
    val expression = "3 + 4 * 2 / (1 - 5) ^ 2 ^ 3"
    println("Postfix: ${shuntingYard(expression)}")
}
